app/services/throttle.rb

```diff
@@ -0,0 +1,259 @@
+  end
+
+  def attempts
+    if IdentityConfig.store.redis_throttle_enabled
+      @redis_attempts.to_i
+    else
+      postgres_throttle.attempts
+    end
+  end
+
+  def throttled?
+    if IdentityConfig.store.redis_throttle_enabled
+      !expired? && maxed?
+    else
+      postgres_throttle.throttled?
+    end
+  end
+
+  def throttled_else_increment?
+    if throttled?
+      true
+    else
+      increment!
+      false
+    end
+  end
+
+  def attempted_at
+    if IdentityConfig.store.redis_throttle_enabled
+      redis_attempted_at
+    else
+      postgres_throttle.attempted_at
+    end
+  end
+
+  def expires_at
+    if IdentityConfig.store.redis_throttle_enabled
+      return Time.zone.now if redis_attempted_at.blank?
+      redis_attempted_at + Throttle.attempt_window_in_minutes(throttle_type).minutes
+    else
+      postgres_throttle.expires_at
+    end
+  end
+
+  def remaining_count
+    return 0 if throttled?
+
+    if IdentityConfig.store.redis_throttle_enabled
+      Throttle.max_attempts(throttle_type) - attempts
+    else
+      postgres_throttle.remaining_count
+    end
+  end
+
+  def redis_attempts
+    return @redis_attempts if defined?(@redis_attempts)
+
+    fetch_state!
+
+    @redis_attempts
+  end
+
+  def redis_attempted_at
+    return @redis_attempted_at if defined?(@redis_attempted_at)
+
+    fetch_state!
+
+    @redis_attempted_at
+  end
+
+  def expired?
+    expires_at <= Time.zone.now
+  end
+
+  def maxed?
+    @redis_attempts && @redis_attempts >= Throttle.max_attempts(throttle_type)
+  end
+
+  def increment!
+    value = nil
+    REDIS_THROTTLE_POOL.with do |client|
+      value, _success = client.multi do |multi|
+        multi.incr(key)
+        multi.expire(
+          key,
+          Throttle.attempt_window_in_minutes(throttle_type).minutes.seconds.to_i,
+        )
+      end
+    end
+
+    @redis_attempts = value.to_i
+    @redis_attempted_at = Time.zone.now
+
+    postgres_throttle.increment
+
+    attempts
+  end
+
+  def fetch_state!
+    value = nil
+    ttl = nil
+    REDIS_THROTTLE_POOL.with do |client|
+      value, ttl = client.multi do |multi|
+        multi.get(key)
+        multi.ttl(key)
+      end
+    end
+
+    @redis_attempts = value.to_i
+
+    if ttl < 0
+      @redis_attempted_at = nil
+    else
+      @redis_attempted_at =
+        Time.zone.now +
+        Throttle.attempt_window_in_minutes(throttle_type).minutes - ttl.seconds
+    end
+
+    self
+  end
+
+  def reset!
+    REDIS_THROTTLE_POOL.with do |client|
+      client.del(key)
+    end
+
+    postgres_throttle.reset
+
+    @redis_attempts = 0
+    @redis_attempted_at = nil
+  end
+
+  def set_as_throttled!
+    value = nil
+    REDIS_THROTTLE_POOL.with do |client|
+      value, _success = client.multi do |multi|
+        multi.set(key, Throttle.max_attempts(throttle_type))
+        multi.expire(
+          key,
+          Throttle.attempt_window_in_minutes(throttle_type).minutes.seconds.to_i,
+        )
+      end
```

Comment:
good candidate for setex? Looks like it wouldn't read `value` back but 🤷 

```suggestion
      value = Throttle.max_attempts(throttle_type)
      client.setex(
        key,
        Throttle.attempt_window_in_minutes(throttle_type).minutes.seconds.to_i,
        value,
      )
```